RFCs/FS-1135-random-functions-for-collections.md

```diff
@@ -0,0 +1,206 @@
+
+This RFC covers the detailed proposal for this suggestion.
+
+- [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/508)
+- [x] Approved in principle
+- [ ] [Implementation]() (no implementation yet)
+- [ ] Design Review Meeting(s) with @dsyme and others invitees
+- [Discussion](https://github.com/fsharp/fslang-design/discussions/731)
+
+# Summary
+
+This feature extends collections apis with functions for random sampling and shuffling built-in fsharp collections.
+
+# Motivation
+
+This feature is motivated by the following use cases:
+ - Using F# for data science and machine learning (like building a neural network), where data shuffling plays important role
+ - Building games, where random sampling is used for generating random levels, random decks, etc.
+ - Building simulations, where random sampling is used for generating random input data
+
+# Detailed design
+
+### General
+
+The following general rules are applied to all functions
+ - New functions should be implemented in List, Array, Seq modules
+ - All functions should have a variant with a [Random](https://learn.microsoft.com/en-us/dotnet/api/system.random) parameter
+ - Shared thread-safe Random instance should be used for all basic functions
+
+### Shuffle
+
+Shuffle function should returned a new shuffled collection of the same collection type.
+
+Two functions should be added to each module.
+
+```fsharp
+// Array module
+val randomShuffle: array:'T[] -> 'T[]
+val randomShuffleWith: random:Random -> array:'T[] -> 'T[]
+val randomShuffleInPlace: array:'T[] -> 'T[]
+val randomShuffleInPlaceWith: random:Random -> array:'T[] -> 'T[]
+// List module
+val randomShuffle: list:'T list -> 'T list
+val randomShuffleWith: random:Random -> list:'T list -> 'T list
+// Seq module
+val randomShuffle: source:'T seq -> 'T seq
+val randomShuffleWith: random:Random -> source:'T seq -> 'T seq
+```
+[ArgumentNullException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentnullexception) should be raised if collection is null
+
+Example:
+```fsharp
+let allPlayers = [ "Alice"; "Bob"; "Charlie"; "Dave" ]
+let round1Order = allPlayers |> List.randomShuffle // [ "Charlie"; "Dave"; "Alice"; "Bob" ]
+```
+
+### Choice
+
+Choice function should returned a random element from the collection.
+
+Two functions should be added to each module.
+
+```fsharp
+// Array module
+val randomChoice: array:'T[] -> 'T
+val randomChoiceWith: random:Random -> array:'T[] -> 'T
+// List module
+val randomChoice: list:'T list -> 'T
+val randomChoiceWith: random:Random -> list:'T list -> 'T
+// Seq module
+val randomChoice: source:'T seq -> 'T
+val randomChoiceWith: random:Random -> source:'T seq -> 'T
+```
+[ArgumentNullException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentnullexception) should be raised if collection is null
```

Comment:
And if the `Random` instance is `null` (depending on decision taken in #731)
```suggestion
[`ArgumentNullException`](https://learn.microsoft.com/en-us/dotnet/api/system.argumentnullexception) is raised if collection argument is `null`, or if the `random` argument is `null`.
```

Reply:
Fixed